### ISSUE TITLE
threading with python bindings

### ISSUE THREAD

**Issue by bogklug** [linux-audit/audit-userspace#305]
Hi,

It seems that python bindings do not really support threading now (I don't see any Py_BEGIN_ALLOW_THREADS, Py_END_ALLOW_THREADS in auparse_python.c and e.g. calling parse_next_event or search_next_event on an AuParser of source type AUSOURCE_FILE_POINTER blocks all python threads).

My usecase is to have one thread reading the events and another sending them via HTTP to some external service. I guess for now the simplest workaround for me might be switching from threading to multiprocessing.

Also, perhaps developing a python async interface for auparse is worth considering as an alternative to threading/multiprocessing.

**Comment by stevegrubb** [comment]
Have you looked into the auparse feed API? It does callbacks when the event is ready. So, you can have a loop collecting records and feeding them. When the event is complete, it hands it to a callback function. Perhaps the callback does formatting and writes it to the service? Also, not too likely be any further development of the python bindings.

**Comment by bogklug** [comment]
Thanks for your suggestion. Yes, the feed API seems to be the simplest solution...

I rewrote my plugin with multiprocessing and then multiprocessing+threading, but this time I had problems with signals/deadlocks (not 100% sure, but I'm suspecting it was EINTR related, i.e. search_next_event stdin read being interrupted by a signal; python closes stdin on child processes when using multiprocessing.Process, so the main process needs to both handle signals and read from stdin (though maybe some hackish workaround is possible)).

In the end I rewrote once again using the feed API and threading and it seems ok this time.
